fem/bilininteg_hcurl.cpp

```diff
@@ -1802,6 +1802,515 @@ void PAHcurlH1Apply3D(const int D1D,
+         {
+            vw[dy][ex] = 0.0;
+            for (int ey = 0; ey < o_dofs1D; ++ey)
+            {
+               const int local_index = c_dofs1D * o_dofs1D + ey*c_dofs1D + ex;
+               vw[dy][ex] += G(ey, dy) * x(local_index, e);
+            }
+         }
+      }
+
+      for (int dy = 0; dy < c_dofs1D; ++dy)
+      {
+         for (int dx = 0; dx < c_dofs1D; ++dx)
+         {
+            for (int ex = 0; ex < c_dofs1D; ++ex)
+            {
+               y(dx, dy, e) += B(ex, dx) * vw[dy][ex];
+            }
+         }
+      }
+   });
+}
+
+static void PAHcurlApplyGradient3D(const int c_dofs1D,
+                                   const int o_dofs1D,
+                                   const int NE,
+                                   const Array<double> &_B,
+                                   const Array<double> &_G,
+                                   const Vector &_x,
+                                   Vector &_y)
+{
+   auto B = Reshape(_B.Read(), c_dofs1D, c_dofs1D);
+   auto G = Reshape(_G.Read(), o_dofs1D, c_dofs1D);
+
+   auto x = Reshape(_x.Read(), c_dofs1D, c_dofs1D, c_dofs1D, NE);
+   auto y = Reshape(_y.ReadWrite(), (3 * c_dofs1D * c_dofs1D * o_dofs1D), NE);
+
+   MFEM_FORALL(e, NE,
+   {
+      double pxw1[c_dofs1D][c_dofs1D][c_dofs1D];
+      double pxw2[c_dofs1D][c_dofs1D][c_dofs1D];
+      double pyw1[c_dofs1D][c_dofs1D][c_dofs1D];
+      double pyw2[c_dofs1D][o_dofs1D][c_dofs1D];
+      double pzw1[c_dofs1D][c_dofs1D][o_dofs1D];
+      double pzw2[c_dofs1D][c_dofs1D][o_dofs1D];
+
+      // ---
+      // dofs that point parallel to x-axis (open in x, closed in y, z)
+      // ---
+
+      // contract in z
+      for (int ez = 0; ez < c_dofs1D; ++ez)
+      {
+         for (int dx = 0; dx < c_dofs1D; ++dx)
+         {
+            for (int dy = 0; dy < c_dofs1D; ++dy)
+            {
+               pxw1[dx][dy][ez] = 0.0;
+               for (int dz = 0; dz < c_dofs1D; ++dz)
+               {
+                  pxw1[dx][dy][ez] += B(ez, dz) * x(dx, dy, dz, e);
+               }
+            }
+         }
+      }
+
+      // contract in y
+      for (int ez = 0; ez < c_dofs1D; ++ez)
+      {
+         for (int ey = 0; ey < c_dofs1D; ++ey)
+         {
+            for (int dx = 0; dx < c_dofs1D; ++dx)
+            {
+               pxw2[dx][ey][ez] = 0.0;
+               for (int dy = 0; dy < c_dofs1D; ++dy)
+               {
+                  pxw2[dx][ey][ez] += B(ey, dy) * pxw1[dx][dy][ez];
+               }
+            }
+         }
+      }
+
+      // contract in x
+      for (int ez = 0; ez < c_dofs1D; ++ez)
+      {
+         for (int ey = 0; ey < c_dofs1D; ++ey)
+         {
+            for (int ex = 0; ex < o_dofs1D; ++ex)
+            {
+               for (int dx = 0; dx < c_dofs1D; ++dx)
+               {
+                  const int local_index = ez*c_dofs1D*o_dofs1D + ey*o_dofs1D + ex;
+                  y(local_index, e) += G(ex, dx) * pxw2[dx][ey][ez];
```

Comment:
In summations like this, I think it is more efficient to use a local variable `double sum = 0.0` and then set `y(local_index, e) = sum` after the `dx` loop.